src/client/Microsoft.Identity.Client/Instance/AuthorityResolutionManager.cs

```diff
@@ -3,54 +3,79 @@
 
 using System.Collections.Concurrent;
 using System.Collections.Generic;
+using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.Identity.Client.Instance.Validation;
 using Microsoft.Identity.Client.Internal;
+using Microsoft.Identity.Client.Utils;
 
 namespace Microsoft.Identity.Client.Instance
 {
     /// <summary>
-    /// Responsible for figuring out authority endpoints
+    /// Responsible for figuring out authority endpoints and for validation
     /// </summary>
-    internal class AuthorityEndpointResolutionManager : IAuthorityEndpointResolutionManager
+    internal class AuthorityResolutionManager
+        : IAuthorityResolutionManager
     {
         private static readonly ConcurrentDictionary<string, AuthorityEndpointCacheEntry> s_endpointCacheEntries =
             new ConcurrentDictionary<string, AuthorityEndpointCacheEntry>();
 
+        private static readonly ConcurrentHashSet<string> s_validatedEnvironments =
+            new ConcurrentHashSet<string>();
+
         private readonly IServiceBundle _serviceBundle;
+        private readonly SemaphoreSlim _semaphoreSlim = new SemaphoreSlim(1, 1);
 
-        public AuthorityEndpointResolutionManager(IServiceBundle serviceBundle, bool shouldClearCache = true)
+        public AuthorityResolutionManager(IServiceBundle serviceBundle, bool shouldClearCache = true)
         {
```

Comment:
i don't think we need the serviceBundle, we don't use it and we can use the RC to get the service bundle anyway, so if we pass the RC instead of the service bundle in IAuthorityValidator, then we don't have to pass the RC later, as we'll have it in the contructor. I made a draft PR w/those changes, as it was easier to review this with the branch anyway. [https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2546](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2546)